tempesta_fw/http.c

```diff
@@ -524,46 +520,14 @@ tfw_h2_prep_redirect(TfwHttpResp *resp, unsigned short status, TfwStr *rmark,
 	 * 'content-length' header also must be added into HEADERS frame.
 	 */
 	if (body) {
-		TfwStr f_hdr = {
-			.data = buf,
-			.len = sizeof(buf)
-		};
-
-		if (WARN_ON_ONCE(!body->len))
-			return -EINVAL;
-
-		if (body->len > FRAME_MAX_LENGTH)
-			return -E2BIG;
-
-		frame_hdr.length = body->len;
-		frame_hdr.type = HTTP2_DATA;
-		frame_hdr.flags = HTTP2_F_END_STREAM;
-		tfw_h2_pack_frame_header(buf, &frame_hdr);
-
-		r = tfw_http_msg_expand_data(iter, skb_head, &f_hdr, NULL);
-		if (unlikely(r))
-			return r;
-
 		r = tfw_http_msg_expand_data(iter, skb_head, body, NULL);
 		if (unlikely(r))
 			return r;
 	}
```

Comment:
It seems that first DATA frame header is written after the body now. However, to write it before body - it seems that the initialization of `tfw_cfg_js_ch->body` during configuration should be changed in order to generate `content-length` HTTP/2 header separately. Looks like `content-length` header and `body` parts should be splitted into two `TfwStr` instances during configuration, and concatenated only during creation of HTTP/1 redirect response.

Reply:
Yeah, good catch! Didn't tested javascript challenge scenario, which is only affected by the issue. Fixed that.